File: results/2026-05-06_173435/tasks/18-secret-rotation-validator/default-opus/generated-code/run_tests.py
#!/usr/bin/env python3
"""Test harness — runs all tests through act and validates output.

Sets up a temp git repo, copies project files, runs act push, captures output,
and asserts on exact expected values. Also validates workflow structure and actionlint.
"""

import json
import os
import re
import shutil
import subprocess
import sys
import tempfile
import textwrap

SCRIPT_DIR = os.path.dirname(os.path.abspath(__file__))
ACT_RESULT_FILE = os.path.join(SCRIPT_DIR, "act-result.txt")

PROJECT_FILES = [
    "secret_rotation_validator.py",
    "test_secret_rotation_validator.py",
    "fixtures/mixed_config.json",
    "fixtures/all_expired_config.json",
    ".github/workflows/secret-rotation-validator.yml",
]


def copy_project_to(dest: str):
    for f in PROJECT_FILES:
        src = os.path.join(SCRIPT_DIR, f)
        dst = os.path.join(dest, f)
        os.makedirs(os.path.dirname(dst), exist_ok=True)
        shutil.copy2(src, dst)
    actrc = os.path.join(dest, ".actrc")
    with open(actrc, "w") as fh:
        fh.write("-P ubuntu-latest=act-ubuntu-pwsh:latest\n")
        fh.write("--pull=false\n")


def run_act_in_temp_repo(label: str) -> tuple[int, str]:
    """Set up a temp git repo, copy project files, run act push, return (exit_code, output)."""
    tmpdir = tempfile.mkdtemp(prefix=f"act_{label}_")
    try:
        copy_project_to(tmpdir)
        subprocess.run(
            ["git", "init"], cwd=tmpdir, capture_output=True, check=True
        )
        subprocess.run(
            ["git", "add", "-A"], cwd=tmpdir, capture_output=True, check=True
        )
        subprocess.run(
            ["git", "commit", "-m", "test"],
            cwd=tmpdir,
            capture_output=True,
            check=True,
            env={**os.environ, "GIT_AUTHOR_NAME": "test", "GIT_AUTHOR_EMAIL": "t@t.co",
                 "GIT_COMMITTER_NAME": "test", "GIT_COMMITTER_EMAIL": "t@t.co"},
        )
        result = subprocess.run(
            ["act", "push", "--rm"],
            cwd=tmpdir,
            capture_output=True,
            text=True,
            timeout=300,
        )
        output = result.stdout + "\n" + result.stderr
        return result.returncode, output
    finally:
        shutil.rmtree(tmpdir, ignore_errors=True)


def extract_section(output: str, start_marker: str, end_marker: str) -> str:
    """Extract text between markers from act output."""
    pattern = re.escape(start_marker) + r"(.*?)" + re.escape(end_marker)
    match = re.search(pattern, output, re.DOTALL)
    if not match:
        return ""
    return match.group(1).strip()


def strip_act_prefixes(text: str) -> str:
    """Remove act's line prefixes (e.g., '[Job/step]   | content') from captured output."""
    lines = []
    for line in text.splitlines():
        cleaned = re.sub(r"^\[.*?\]\s+\|\s?", "", line)
        lines.append(cleaned)
    return "\n".join(lines)


def main():
    failures = []
    act_output_parts = []

    # --- Workflow structure tests (no act needed) ---
    print("=" * 60)
    print("WORKFLOW STRUCTURE TESTS")
    print("=" * 60)

    wf_path = os.path.join(SCRIPT_DIR, ".github", "workflows", "secret-rotation-validator.yml")

    # Test: workflow file exists
    if not os.path.isfile(wf_path):
        failures.append("STRUCTURE: workflow file does not exist")
        print("FAIL: workflow file does not exist")
    else:
        print("PASS: workflow file exists")

    # Test: valid YAML with expected structure
    try:
        import yaml
        with open(wf_path) as f:
            wf = yaml.safe_load(f)
        triggers = wf.get(True) or wf.get("on", {})
        assert "push" in triggers, "missing push trigger"
        assert "jobs" in wf, "missing jobs"
        assert "validate" in wf["jobs"], "missing validate job"
        step_uses = [s.get("uses", "") for s in wf["jobs"]["validate"]["steps"]]
        assert any("actions/checkout" in u for u in step_uses), "missing checkout step"
        step_runs = " ".join(s.get("run", "") for s in wf["jobs"]["validate"]["steps"])
        assert "secret_rotation_validator.py" in step_runs, "workflow doesn't reference script"
        print("PASS: workflow YAML structure is correct")
    except Exception as e:
        failures.append(f"STRUCTURE: {e}")
        print(f"FAIL: {e}")

    # Test: referenced script files exist
    for fname in ["secret_rotation_validator.py", "fixtures/mixed_config.json", "fixtures/all_expired_config.json"]:
        fpath = os.path.join(SCRIPT_DIR, fname)
        if os.path.isfile(fpath):
            print(f"PASS: referenced file exists: {fname}")
        else:
            failures.append(f"STRUCTURE: referenced file missing: {fname}")
            print(f"FAIL: referenced file missing: {fname}")

    # Test: actionlint passes
    lint_result = subprocess.run(
        ["actionlint", wf_path], capture_output=True, text=True
    )
    if lint_result.returncode == 0:
        print("PASS: actionlint passes")
    else:
        failures.append(f"STRUCTURE: actionlint failed: {lint_result.stdout}{lint_result.stderr}")
        print(f"FAIL: actionlint failed:\n{lint_result.stdout}{lint_result.stderr}")

    # --- Act execution test ---
    print()
    print("=" * 60)
    print("ACT EXECUTION TEST: mixed config (all formats)")
    print("=" * 60)

    exit_code, output = run_act_in_temp_repo("mixed")
    act_output_parts.append(f"{'=' * 60}\nTEST CASE: mixed config\n{'=' * 60}\n{output}")

    # Assert exit code 0
    if exit_code == 0:
        print("PASS: act exited with code 0")
    else:
        failures.append(f"ACT: exit code was {exit_code}, expected 0")
        print(f"FAIL: act exited with code {exit_code}")

    # Assert job succeeded (act uses either "Job succeeded" or emoji markers)
    if "Job succeeded" in output:
        print("PASS: Job succeeded found in output")
    elif exit_code == 0:
        print("PASS: Job succeeded (exit code 0)")
    else:
        failures.append("ACT: job did not succeed")
        print("FAIL: job did not succeed")

    # Extract and validate JSON mixed output
    json_section = strip_act_prefixes(extract_section(output, "===JSON_MIXED_START===", "===JSON_MIXED_END==="))
    if json_section:
        try:
            data = json.loads(json_section)
            checks = [
                (data["reference_date"] == "2026-05-07", "reference_date is 2026-05-07"),
                (data["warning_window_days"] == 14, "warning_window_days is 14"),
                (data["summary"]["total"] == 5, "total is 5"),
                (data["summary"]["expired"] == 2, "expired count is 2"),
                (data["summary"]["warning"] == 1, "warning count is 1"),
                (data["summary"]["ok"] == 2, "ok count is 2"),
                ([s["name"] for s in data["secrets"]["expired"]] == ["DB_PASSWORD", "OAUTH_SECRET"],
                 "expired secrets are DB_PASSWORD, OAUTH_SECRET"),
                ([s["name"] for s in data["secrets"]["warning"]] == ["TLS_CERT"],
                 "warning secret is TLS_CERT"),
                ([s["name"] for s in data["secrets"]["ok"]] == ["API_KEY", "SSH_KEY"],
                 "ok secrets are API_KEY, SSH_KEY"),
                (data["secrets"]["expired"][0]["days_since_rotation"] == 157,
                 "DB_PASSWORD days_since_rotation is 157"),
                (data["secrets"]["expired"][0]["days_until_expiry"] == -67,
                 "DB_PASSWORD days_until_expiry is -67"),
                (data["secrets"]["warning"][0]["days_until_expiry"] == 7,
                 "TLS_CERT days_until_expiry is 7"),
                (data["secrets"]["ok"][0]["days_until_expiry"] == 18,
                 "API_KEY days_until_expiry is 18"),
            ]
            for ok, desc in checks:
                if ok:
                    print(f"PASS: JSON mixed - {desc}")
                else:
                    failures.append(f"ACT JSON: {desc}")
                    print(f"FAIL: JSON mixed - {desc}")
        except json.JSONDecodeError as e:
            failures.append(f"ACT JSON: could not parse JSON output: {e}")
            print(f"FAIL: could not parse JSON output: {e}")
    else:
        failures.append("ACT JSON: JSON mixed section not found in output")
        print("FAIL: JSON mixed section not found in output")

    # Validate markdown output
    md_section = strip_act_prefixes(extract_section(output, "===MARKDOWN_MIXED_START===", "===MARKDOWN_MIXED_END==="))
    if md_section:
        md_checks = [
            ("# Secret Rotation Report" in md_section, "has title header"),
            ("| Expired | 2 |" in md_section, "expired count 2 in summary"),
            ("| Warning | 1 |" in md_section, "warning count 1 in summary"),
            ("| OK | 2 |" in md_section, "ok count 2 in summary"),
            ("| **Total** | **5** |" in md_section, "total 5 in summary"),
            ("DB_PASSWORD" in md_section, "contains DB_PASSWORD"),
            ("TLS_CERT" in md_section, "contains TLS_CERT"),
            ("API_KEY" in md_section, "contains API_KEY"),
            ("67" in md_section, "contains days overdue 67"),
        ]
        for ok, desc in md_checks:
            if ok:
                print(f"PASS: Markdown - {desc}")
            else:
                failures.append(f"ACT Markdown: {desc}")
                print(f"FAIL: Markdown - {desc}")
    else:
        failures.append("ACT Markdown: markdown section not found in output")
        print("FAIL: markdown section not found in output")

    # Validate all-expired config output
    expired_section = strip_act_prefixes(
        extract_section(output, "===JSON_EXPIRED_START===", "===JSON_EXPIRED_END===")
    )
    if expired_section:
        try:
            data = json.loads(expired_section)
            checks = [
                (data["summary"]["expired"] == 3, "all 3 secrets are expired"),
                (data["summary"]["warning"] == 0, "no warning secrets"),
                (data["summary"]["ok"] == 0, "no ok secrets"),
                ([s["name"] for s in data["secrets"]["expired"]] == ["OLD_DB_PASS", "OLD_API_KEY", "OLD_CERT"],
                 "expired secrets sorted by urgency"),
            ]
            for ok, desc in checks:
                if ok:
                    print(f"PASS: JSON expired - {desc}")
                else:
                    failures.append(f"ACT expired: {desc}")
                    print(f"FAIL: JSON expired - {desc}")
        except json.JSONDecodeError as e:
            failures.append(f"ACT expired: could not parse JSON: {e}")
            print(f"FAIL: could not parse all-expired JSON: {e}")
    else:
        failures.append("ACT expired: section not found in output")
        print("FAIL: all-expired section not found in output")

    # Validate warning window override output
    override_section = strip_act_prefixes(
        extract_section(output, "===JSON_OVERRIDE_START===", "===JSON_OVERRIDE_END===")
    )
    if override_section:
        try:
            data = json.loads(override_section)
            checks = [
                (data["warning_window_days"] == 20, "warning window is 20"),
                (data["summary"]["warning"] == 2, "2 warnings with 20-day window"),
                (data["summary"]["ok"] == 1, "1 ok with 20-day window"),
            ]
            for ok, desc in checks:
                if ok:
                    print(f"PASS: JSON override - {desc}")
                else:
                    failures.append(f"ACT override: {desc}")
                    print(f"FAIL: JSON override - {desc}")
        except json.JSONDecodeError as e:
            failures.append(f"ACT override: could not parse JSON: {e}")
            print(f"FAIL: could not parse override JSON: {e}")
    else:
        failures.append("ACT override: section not found in output")
        print("FAIL: override section not found in output")

    # Check pytest ran and passed (allow skipped tests, but no failures)
    pytest_summary = re.search(r"(\d+) passed", output)
    pytest_failed = re.search(r"(\d+) failed", output)
    if pytest_summary and (not pytest_failed or int(pytest_failed.group(1)) == 0):
        print(f"PASS: pytest passed in workflow ({pytest_summary.group(0)})")
    else:
        failures.append("ACT: pytest did not pass cleanly")
        print("FAIL: pytest did not pass cleanly in workflow")

    # --- Write act-result.txt ---
    with open(ACT_RESULT_FILE, "w") as f:
        f.write("\n\n".join(act_output_parts))
    print(f"\nWrote act output to {ACT_RESULT_FILE}")

    # --- Summary ---
    print()
    print("=" * 60)
    if failures:
        print(f"FAILED: {len(failures)} assertion(s) failed:")
        for fail in failures:
            print(f"  - {fail}")
        sys.exit(1)
    else:
        print("ALL ASSERTIONS PASSED")
        sys.exit(0)


if __name__ == "__main__":
    main()

File: results/2026-05-06_173435/tasks/18-secret-rotation-validator/typescript-bun-opus/generated-code/run-tests.ts
import { existsSync, mkdtempSync, readFileSync, writeFileSync, cpSync, mkdirSync, rmSync } from "fs";
import { join, resolve } from "path";
import { tmpdir } from "os";
import { spawnSync } from "child_process";

const projectDir = resolve(import.meta.dir);
const resultLines: string[] = [];
let failCount = 0;
let passCount = 0;

function log(msg: string): void {
  console.log(msg);
  resultLines.push(msg);
}

function assert(condition: boolean, message: string): void {
  if (!condition) {
    log(`  FAIL: ${message}`);
    failCount++;
  } else {
    log(`  PASS: ${message}`);
    passCount++;
  }
}

function extractBetweenMarkers(output: string, startMarker: string, endMarker: string): string {
  const lines = output.split("\n");
  let capturing = false;
  const captured: string[] = [];
  for (const line of lines) {
    if (line.includes(startMarker)) {
      capturing = true;
      continue;
    }
    if (line.includes(endMarker)) {
      capturing = false;
      continue;
    }
    if (capturing) {
      const pipeMatch = line.match(/\|\s?(.*)/);
      if (pipeMatch) {
        captured.push(pipeMatch[1]);
      }
    }
  }
  return captured.join("\n");
}

// ── Phase 1: Workflow structure tests ──
log("=== PHASE 1: WORKFLOW STRUCTURE TESTS ===");

const workflowPath = join(projectDir, ".github/workflows/secret-rotation-validator.yml");
assert(existsSync(workflowPath), "Workflow file exists at .github/workflows/secret-rotation-validator.yml");

const wf = readFileSync(workflowPath, "utf-8");
assert(wf.includes("on:"), "Workflow has 'on:' trigger block");
assert(wf.includes("push"), "Workflow triggers on push");
assert(wf.includes("pull_request"), "Workflow triggers on pull_request");
assert(wf.includes("workflow_dispatch"), "Workflow has workflow_dispatch trigger");
assert(wf.includes("schedule"), "Workflow has schedule trigger");
assert(/cron:/.test(wf), "Workflow has cron expression");
assert(wf.includes("jobs:"), "Workflow has jobs section");
assert(wf.includes("runs-on:"), "Workflow specifies runs-on");
assert(wf.includes("actions/checkout@v4"), "Workflow uses actions/checkout@v4");
assert(wf.includes("bun test"), "Workflow runs bun test");
assert(wf.includes("bun run main.ts"), "Workflow references main.ts script");
assert(wf.includes("--config fixtures/test-secrets.json"), "Workflow references fixtures/test-secrets.json");
assert(wf.includes("--format json"), "Workflow runs JSON format output");
assert(wf.includes("--format markdown"), "Workflow runs Markdown format output");

assert(existsSync(join(projectDir, "main.ts")), "main.ts exists");
assert(existsSync(join(projectDir, "validator.ts")), "validator.ts exists");
assert(existsSync(join(projectDir, "formatter.ts")), "formatter.ts exists");
assert(existsSync(join(projectDir, "types.ts")), "types.ts exists");
assert(existsSync(join(projectDir, "fixtures/test-secrets.json")), "fixtures/test-secrets.json exists");
assert(existsSync(join(projectDir, "validator.test.ts")), "validator.test.ts exists");
assert(existsSync(join(projectDir, "formatter.test.ts")), "formatter.test.ts exists");

// ── Phase 2: actionlint validation ──
log("\n=== PHASE 2: ACTIONLINT VALIDATION ===");

const lintResult = spawnSync("actionlint", [workflowPath], { encoding: "utf-8" });
assert(
  lintResult.status === 0,
  `actionlint exits with code 0 (got ${lintResult.status})`
);
if (lintResult.status !== 0) {
  log(`  actionlint output:\n${lintResult.stdout}\n${lintResult.stderr}`);
}

// ── Phase 3: act integration tests ──
log("\n=== PHASE 3: ACT INTEGRATION TESTS ===");

const tmpDir = mkdtempSync(join(tmpdir(), "secret-rotation-act-"));
log(`  Temp directory: ${tmpDir}`);

const filesToCopy = [
  "types.ts",
  "validator.ts",
  "formatter.ts",
  "main.ts",
  "validator.test.ts",
  "formatter.test.ts",
  "fixtures/test-secrets.json",
  ".github/workflows/secret-rotation-validator.yml",
];

for (const file of filesToCopy) {
  const src = join(projectDir, file);
  const dest = join(tmpDir, file);
  mkdirSync(join(dest, ".."), { recursive: true });
  cpSync(src, dest);
}

writeFileSync(join(tmpDir, ".actrc"), "-P ubuntu-latest=act-ubuntu-pwsh:latest\n");

const gitEnv = {
  ...process.env,
  GIT_AUTHOR_NAME: "test",
  GIT_COMMITTER_NAME: "test",
  GIT_AUTHOR_EMAIL: "test@test.com",
  GIT_COMMITTER_EMAIL: "test@test.com",
};

spawnSync("git", ["init"], { cwd: tmpDir, env: gitEnv });
spawnSync("git", ["add", "."], { cwd: tmpDir, env: gitEnv });
spawnSync("git", ["commit", "-m", "initial"], { cwd: tmpDir, env: gitEnv });

log("  Running act push --rm (this may take a few minutes)...");
const actResult = spawnSync("act", ["push", "--rm", "--pull=false"], {
  cwd: tmpDir,
  encoding: "utf-8",
  timeout: 300_000,
});

const actOutput = (actResult.stdout || "") + "\n" + (actResult.stderr || "");

log(`  act exit code: ${actResult.status}`);

const actResultPath = join(projectDir, "act-result.txt");
writeFileSync(
  actResultPath,
  `=== ACT RUN: Secret Rotation Validator ===\nExit code: ${actResult.status}\n\n${actOutput}\n`
);
log(`  Saved act output to act-result.txt (${actOutput.length} chars)`);

assert(actResult.status === 0, "act push exits with code 0");

assert(actOutput.includes("Job succeeded"), "act output shows 'Job succeeded'");

// Check bun test results are visible
assert(actOutput.includes("pass"), "act output shows tests passing");
assert(!/[1-9]\d* fail/.test(actOutput), "act output shows no test failures");

// ── Parse JSON output ──
log("\n=== PHASE 4: JSON OUTPUT ASSERTIONS ===");

const jsonContent = extractBetweenMarkers(actOutput, "JSON_OUTPUT_START", "JSON_OUTPUT_END");
let jsonParsed = false;

if (jsonContent.trim()) {
  try {
    const report = JSON.parse(jsonContent);
    jsonParsed = true;

    assert(report.generatedAt === "2026-05-07", "JSON generatedAt is '2026-05-07'");
    assert(report.warningWindowDays === 14, "JSON warningWindowDays is 14");

    assert(report.summary.total === 4, "JSON summary.total is 4");
    assert(report.summary.expired === 1, "JSON summary.expired is 1");
    assert(report.summary.warning === 1, "JSON summary.warning is 1");
    assert(report.summary.ok === 2, "JSON summary.ok is 2");

    const db = report.secrets.find((s: any) => s.name === "DB_PASSWORD");
    assert(db !== undefined, "JSON contains DB_PASSWORD secret");
    assert(db?.urgency === "expired", "JSON DB_PASSWORD urgency is 'expired'");
    assert(db?.daysSinceRotation === 843, "JSON DB_PASSWORD daysSinceRotation is 843");
    assert(db?.daysUntilExpiry === -753, "JSON DB_PASSWORD daysUntilExpiry is -753");
    assert(db?.expiryDate === "2024-04-14", "JSON DB_PASSWORD expiryDate is '2024-04-14'");

    const jwt = report.secrets.find((s: any) => s.name === "JWT_SECRET");
    assert(jwt !== undefined, "JSON contains JWT_SECRET secret");
    assert(jwt?.urgency === "warning", "JSON JWT_SECRET urgency is 'warning'");
    assert(jwt?.daysUntilExpiry === 13, "JSON JWT_SECRET daysUntilExpiry is 13");
    assert(jwt?.expiryDate === "2026-05-20", "JSON JWT_SECRET expiryDate is '2026-05-20'");

    const api = report.secrets.find((s: any) => s.name === "API_KEY");
    assert(api !== undefined, "JSON contains API_KEY secret");
    assert(api?.urgency === "ok", "JSON API_KEY urgency is 'ok'");
    assert(api?.daysUntilExpiry === 208, "JSON API_KEY daysUntilExpiry is 208");

    const slack = report.secrets.find((s: any) => s.name === "SLACK_WEBHOOK");
    assert(slack !== undefined, "JSON contains SLACK_WEBHOOK secret");
    assert(slack?.urgency === "ok", "JSON SLACK_WEBHOOK urgency is 'ok'");
    assert(slack?.daysUntilExpiry === 174, "JSON SLACK_WEBHOOK daysUntilExpiry is 174");

    // Check sort order: expired, warning, ok
    assert(
      report.secrets[0].urgency === "expired" &&
        report.secrets[1].urgency === "warning" &&
        report.secrets[2].urgency === "ok" &&
        report.secrets[3].urgency === "ok",
      "JSON secrets sorted by urgency (expired, warning, ok)"
    );
  } catch (e: any) {
    assert(false, `JSON output is valid JSON: ${e.message}`);
  }
} else {
  assert(false, "JSON output found between markers in act output");
}

assert(jsonParsed, "JSON output was successfully parsed and validated");

// ── Parse Markdown output ──
log("\n=== PHASE 5: MARKDOWN OUTPUT ASSERTIONS ===");

const mdContent = extractBetweenMarkers(actOutput, "MARKDOWN_OUTPUT_START", "MARKDOWN_OUTPUT_END");

assert(mdContent.includes("Secret Rotation Report"), "Markdown contains 'Secret Rotation Report' title");
assert(
  mdContent.includes("Generated: 2026-05-07 | Warning Window: 14 days"),
  "Markdown contains generation metadata line"
);
assert(mdContent.includes("| Expired | 1 |"), "Markdown summary shows Expired = 1");
assert(mdContent.includes("| Warning | 1 |"), "Markdown summary shows Warning = 1");
assert(mdContent.includes("| OK | 2 |"), "Markdown summary shows OK = 2");
assert(mdContent.includes("| **Total** | **4** |"), "Markdown summary shows Total = 4");

assert(mdContent.includes("## Expired"), "Markdown has Expired section");
assert(
  mdContent.includes("| DB_PASSWORD | 2024-01-15 | 90 | 2024-04-14 | 753 | api-server, worker |"),
  "Markdown Expired row for DB_PASSWORD with 753 days overdue"
);

assert(mdContent.includes("## Warning"), "Markdown has Warning section");
assert(
  mdContent.includes("| JWT_SECRET | 2026-04-20 | 30 | 2026-05-20 | 13 | auth-service |"),
  "Markdown Warning row for JWT_SECRET with 13 days until expiry"
);

assert(mdContent.includes("## OK"), "Markdown has OK section");
assert(
  mdContent.includes("| API_KEY | 2025-12-01 | 365 | 2026-12-01 | 208 | frontend |"),
  "Markdown OK row for API_KEY with 208 days until expiry"
);
assert(
  mdContent.includes("| SLACK_WEBHOOK | 2026-05-01 | 180 | 2026-10-28 | 174 | notification-service |"),
  "Markdown OK row for SLACK_WEBHOOK with 174 days until expiry"
);

// ── Cleanup ──
rmSync(tmpDir, { recursive: true, force: true });

// ── Final summary ──
log("\n========================================");
log(`  Total: ${passCount + failCount} | Passed: ${passCount} | Failed: ${failCount}`);
log("========================================");

if (failCount > 0) {
  writeFileSync(actResultPath, readFileSync(actResultPath, "utf-8") + "\n\n" + resultLines.join("\n") + "\n");
  process.exit(1);
} else {
  writeFileSync(actResultPath, readFileSync(actResultPath, "utf-8") + "\n\n" + resultLines.join("\n") + "\n");
  log("All tests passed!");
}
